### ISSUE TITLE
SVG Pins: performance issues

### ISSUE THREAD

**Issue by janusw** [Mapsui/Mapsui#1084]
I suspect that the memory handling for SVG Pins is not quite optimal, and might have some issues.

Let me illustrate this with the following scenario:
* I have a map based on Mapsui.UI.Forms.MapView with a potentially larger number of pins.
* There are a number of different types of pins, where each type has its own image (e.g. given by an svg or png file). Let's say we have something like 10-20 types, and each image file has a few kB.
* The map will contain many pins of each type (say: hundreds or thousands).

Since there are many pins, I'd like to avoid duplicating the memory allocated for their images, for reasons of performance and memory footprint. I can see some potential issues in this direction (I've been looking at the `Pin` class and `Pin.CreateFeature` in particular):

1. For svg pins, the svg data is stored as a string (`Pin.Svg`). In `Pin.CreateFeature` this is then converted (copied) into a byte array via `Encoding.UTF8.GetBytes`. This means that for every Pin object I create, I make a separate copy of my svg data, right? For small svgs this might not be a problem, for for more complex ones that have several kB in size, it adds up. And it seems to be avoidable by making Pin.Svg a byte array (just like Pin.Icon) instead of a string. Correct?

2. IIUC the `BitmapRegistry` class could be used to avoid duplication of image data, but it seems to me that for every new pin a new entry in the BitmapRegistry is generated (since `Pin.CreateFeature` alwas seems to call `BitmapRegistry.Instance.Register`). Is that right, or have I simply not looked deep enough into the code? For png pins it might not be a problem: We might have multiple registry entries, but ultimately they all point to the same data in memory. But for svg (since we make a copy along the way, see first point above), we have multiple entries and they all point to their own private copy of the data, right?

@pauldendulk @charlenni 
I'd be grateful for any kinds of comments on this. Am I missing something here?

**Comment by charlenni** [comment]
No, you are correct. `BitmapRegistry` doesn't check, if a "Bitmap" is already registered. And the Forms implementation is very dump in this case too. I didn't implement a check, if the SVG string is already registered. As always: MapView is a convenience wrapper around MapControl for Forms. If you want to do more than the basics of Mapsui, then you should use MapControl. There you could check, if your SVG is registered, and if yes, then use the ID again.

The conversion to a byte array was done, because `BitmapRegistry` once upon a time only accepted this format.

**Comment by janusw** [comment]
> No, you are correct. `BitmapRegistry` doesn't check, if a "Bitmap" is already registered. And the Forms implementation is very dump in this case too. I didn't implement a check, if the SVG string is already registered. As always: MapView is a convenience wrapper around MapControl for Forms. If you want to do more than the basics of Mapsui, then you should use MapControl. There you could check, if your SVG is registered, and if yes, then use the ID again.

Thanks for the comments, @charlenni.

I just stumbled upon ...
https://github.com/Mapsui/Mapsui/blob/8c527ea79d8153f0e448dcf0064d5647ae01bef9/Samples/Mapsui.Samples.Common/Maps/SvgSample.cs#L73
... which seems to implement something like this, right? However, I'm not sure if this approach is feasible with a `Pin`, since the `_bitmapId` is private there, and not accessible from the outside, right?

Could you be a bit more specific about how to check if the SVG is registered already?


> The conversion to a byte array was done, because `BitmapRegistry` once upon a time only accepted this format.

So, are you saying that the conversion is not necessary any more? If so, what's the best way to get rid of it of?

From my point of view, it would be ideal to set the svg data via a stream, not a string. After all, a stream is what is put into the bitmap registry, and I also get the SVG data from the assembly as a stream (via `Assembly.GetManifestResourceStream`).

Instead of reading a string from it, which is then converted into a byte array, from which we again create a MemoryStream, couldn't we directly pass the original stream?

**Comment by janusw** [comment]
Btw, when looking through the list of samples, there are two that are closely related to the issue raised here:
* Mapsui.Samples.Forms.ManyPinsSample
* Mapsui.Samples.Common.Maps.SvgSample

The first one is based on Mapsui.UI.Forms, while the second is not. Both present basically a static map with a large number of pins/symbols (1000 and 2000, respectively). In the first case, the class Mapsui.UI.Forms.Pin is used (of type PinType.Pin), in the second case the class Mapsui.Providers.Feature, with a SymbolStyle based on an SVG image.

Interestingly the ManyPinsSample seems to be much more responsive and performant than the SvgSample, even when using the same number of points. I observe this when running them via Mapsui.Samples.Forms.iOS on an iPhone 8. When zooming out as much as possible and then dragging the map with one finger, the ManyPinsSample follows very fluently, while the SvgSample has large gaps and reacts quite slow.

If anything, I'd have expected the opposite (due to SvgSample appearing to be more low-level, and due to the optimization mentioned above (https://github.com/Mapsui/Mapsui/issues/1084#issuecomment-786655954). Can anyone tell me why the ManyPinsSample is so much more performant? (And can anyone even reproduce this behavior?)

**Comment by janusw** [comment]
> If anything, I'd have expected the opposite (due to SvgSample appearing to be more low-level, and due to the optimization mentioned above ([#1084 (comment)](https://github.com/Mapsui/Mapsui/issues/1084#issuecomment-786655954)). Can anyone tell me why the ManyPinsSample is so much more performant? (And can anyone even reproduce this behavior?)

Btw, the imageCache optimization was apparently implemented by @inforithmics in 5034bc6d551eaf573e78b42982510ade9e97fcae. @inforithmics, do you have any further insights here?

**Comment by janusw** [comment]
I tried some modifications of Mapsui.Samples.Forms.ManyPinsSample:

1. Using PinType.Icon with
`Icon = assembly.GetManifestResourceStream("Mapsui.Samples.Common.Images.loc.png").ToBytes();`
results in a map that is just as responsive as the original one (the png file is 3kB). Check.

2. Using PinType.Svg with the following modifications ...
```
-- Samples/Mapsui.Samples.Forms/Mapsui.Samples.Forms.Shared/ManyPinsSample.cs --
index 376a8aa8..66f10799 100644
@@ -129,12 +129,18 @@ namespace Mapsui.Samples.Forms
         {
             var position = new Position(0 + rnd.Next(-85000, +85000) / 1000.0, 0 + rnd.Next(-180000, +180000) / 1000.0);
 
+            var assembly = typeof(AllSamples).GetTypeInfo().Assembly;
+            var stream = assembly.GetManifestResourceStream("Mapsui.Samples.Common.Images.Pin.svg");
+            StreamReader reader = new StreamReader(stream);
+            string svgString = reader.ReadToEnd();
+
             var pin = new Pin()
             {
-                Label = $"PinType.Pin {num++}",
+                Label = $"PinType.Svg {num++}",
                 Address = position.ToString(),
                 Position = position,
-                Type = PinType.Pin,
+                Type = PinType.Svg,
+                Svg = svgString,
                 Color = new Xamarin.Forms.Color(rnd.Next(0, 256) / 256.0, rnd.Next(0, 256) / 256.0, rnd.Next(0, 256) / 256.0),
                 Transparency = 0.5f,
                 Scale = rnd.Next(50, 130) / 100f,
```
results in the map becoming very sluggish when zooming out completely. Note that the svg file is only about 300 bytes, i.e. a factor of ten smaller than the png used in step one. Note further that this is the same svg as used internally for PinType.pin in the original version of ManyPinsSample.

I guess this is a good test case for the problem reported in this issue. It shows clearly that there is an issue with PinType.Svg.

**Comment by charlenni** [comment]
If you look at the code from Pin (see [here](https://github.com/Mapsui/Mapsui/blob/b4df2b9388cc94d61f54a09fea7f99967a392e54/Mapsui.UI.Forms/Objects/Pin.cs#L459)), then you could see, that, if you use Pin as PinType, the SVG drawing is converted to a bitmap. This is done, because you could then change the color. I didn't find a possibility to do this for SVGs. That's the reason, why these pins are not so clear as the pins in the SVG sample.

So, if you use PintType.SVG, you get an SVG pin. If you use a PinType.Pin, you get a bitmap pin. The later one is drawn much faster, because only a bitmap has to be copied to a place on the canvas, while by the first an SVG image has to be interpreted and then be drawn on the canvas. That's it.

**Comment by janusw** [comment]
> If you look at the code from Pin (see [here](https://github.com/Mapsui/Mapsui/blob/b4df2b9388cc94d61f54a09fea7f99967a392e54/Mapsui.UI.Forms/Objects/Pin.cs#L459)), then you could see, that, if you use Pin as PinType, the SVG drawing is converted to a bitmap.

Yeah, already saw that.


> So, if you use PintType.SVG, you get an SVG pin. If you use a PinType.Pin, you get a bitmap pin. The later one is drawn much faster, because only a bitmap has to be copied to a place on the canvas, while by the first an SVG image has to be interpreted and then be drawn on the canvas. That's it.

Ok, so the difference in behavior of the samples is not related to memory issues at all, but just due to the fact that an svg renders much slower than a bitmap. I guess that means if I need a responsive map with many pins, I better avoid SVG completely. Thanks for clearing that up.

So then, with this in mind, wouldn't it be a good idea to add a new pin type (maybe called `PinType.CachedSvg`), which would take SVG data as input, but cache it into a bitmap for performance reasons (as for `PinType.Pin`, but with a freely chosen svg as input)? If you agree, I'd volunteer to implement that.

Apart from this, I'd argue that `Pin.Svg` should be `byte[]` instead of `string` (analogous to `Pin.Icon`). This makes it easier to use for typical cases, is less confusing and avoids extra conversion. It's a breaking change to the API, though, but since Mapsui 3 is in alpha state, I hope that's fine. If there are no objections, I'll prepare a patch for this as well.

**Comment by charlenni** [comment]
Instead of PinType.CachedSVG you could use PinType.Icon for this. If I would see a PinType.CachedSVG, I would expect, that each time an SVG is drawn. But this isn't the case. So, create your own bitmap from an SVG and use this as icon.

Where do you see these conversions from string to byte array to stream?

And be careful changing anything in this part. You have to remember, that you have to change the same in the Xaml renderer.

**Comment by janusw** [comment]
> Instead of PinType.CachedSVG you could use PinType.Icon for this. If I would see a PinType.CachedSVG, I would expect, that each time an SVG is drawn. But this isn't the case. So, create your own bitmap from an SVG and use this as icon.

Well, of course I can do the conversion myself and use PinType.Icon. But I guess it would be a convenient option for people, if one could simply use an svg as input, and cache it into a bitmap internally to get better performance (in particular since the necessary code is basically already there and could be shared with PinType.Pin, which is also nothing but a convenience option).

Expectations can vary of course. So, if you see `PinType.CachedSVG`, there will be documentation that will tell you what it does. This documentation should also mention the performance improvement over `PinType.SVG` and the slight loss of quality. This will give users the choice of performance vs quality, and keep naive guys like me from falling into this performance trap ;)

> Where do you see these conversions from string to byte array to stream?

Here:

https://github.com/Mapsui/Mapsui/blob/b4df2b9388cc94d61f54a09fea7f99967a392e54/Mapsui.UI.Forms/Objects/Pin.cs#L454

and in my user-side code I have the inverse (stream -> string). I opened #1085 now for using a byte array. Any comments to this?


> And be careful changing anything in this part. You have to remember, that you have to change the same in the Xaml renderer.

I don't quite understand this comment. You mean regarding the new PinType?

**Comment by janusw** [comment]
@pauldendulk, what do you think about my idea with `PinType.CachedSVG`? Maybe it just needs a better name?

**Comment by charlenni** [comment]
Wouldn't it be better to create a helper function, that converts an SVG to an Icon and then use this? Or create a new PinType like PinType.Own, where you could provide your own BitmapID? With the last you could register your few bitmaps and use it for more than one pin.

**Comment by janusw** [comment]
> Wouldn't it be better to create a helper function, that converts an SVG to an Icon and then use this?

In which way is that better?

> Or create a new PinType like PinType.Own, where you could provide your own BitmapID? With the last you could register your few bitmaps and use it for more than one pin.

Actually I don't fully understand what's the point of using this `BitmapRegistry` at all, if it does not even help to deduplicate the bitmap references. Why not use references to the bitmap data directly instead of generating integer IDs for them?

Moreover it seems that the `BitmapRegistry` is somewhat misnamed, since you put not only bitmap data into it, but also SVG vector data.

**Comment by pauldendulk** [comment]
A quick reaction without fully reading this thread. 

This is a very complicated topic. Some smart form of caching may seem the best solution, but there are many different scenarios in which Mapsui is used, which need different caching strategies. What I would like to avoid is some smart solution with a 'we know what is best for you' mentality. I would prefer a solution where it is clear to the developers what is happening so they can use this understanding to develop a solution that is suited for their use case.

**Comment by inforithmics** [comment]
> > If anything, I'd have expected the opposite (due to SvgSample appearing to be more low-level, and due to the optimization mentioned above ([#1084 (comment)](https://github.com/Mapsui/Mapsui/issues/1084#issuecomment-786655954)). Can anyone tell me why the ManyPinsSample is so much more performant? (And can anyone even reproduce this behavior?)
> 
> Btw, the imageCache optimization was apparently implemented by @inforithmics in [5034bc6](https://github.com/Mapsui/Mapsui/commit/5034bc6d551eaf573e78b42982510ade9e97fcae). @inforithmics, do you have any further insights here?

1. When I migrated to skia.svg i discovered that the ManyPinsSample was very slow, the main reason was that every pin got its own BitmapId even if it was the same Image. So 1000 pins meant 1000 MemoryStreams where loaded. So to only have one Memory Stream I implemented caching so the same svg image gets the same Id. Saving a lot of memory and improving the Performance considerable. Because the MemoryStream was later used for caching having the same memory Stream for the Same made the caching work.

2. Skia.Svg has two parts for painting an image.
    First it Parses the svg into SKSvg (which is much slower than drawing the Image).
    Second it Draws it using SKSvg which is very fast only Vector Draw lines.
    So this means that when I assign the same BitmapId for the Same Svg it is caching the parsed SkSvg and draws it very quickly. 
    
  Important parts to lock at it 
  BitmapHelper (Maybe not the best name, would be better image helper) stores the parse SKSvg Image
```
if (bitmapStream is Stream stream)
            {
                if (stream.IsSvg())
                {
                    var svg = new SKSvg();
                    svg.Load(stream);

                    return new BitmapInfo {Svg = svg};
                }
```
Then there is the SymbolCache where each Bitmap Id is Cached with the corresponding BitmapInfo

  public BitmapInfo GetOrCreate(int bitmapId)
```
        {
            if (_cache.Keys.Contains(bitmapId)) return _cache[bitmapId];
            return _cache[bitmapId] = BitmapHelper.LoadBitmap(BitmapRegistry.Instance.Get(bitmapId));
        }
```

So in conculsion when you do it right. only register one bitmap id per svg pin Kind. Your Svg is cached. if you assign a bitmap id per new pin it isn't cached or you get 1000 svg instances for 1000 pins. Which destroys Performance. 

I will have a look at the SvgSample and report back why it is slower, but I guess that it assigns for every Image a new BitmapId instead a BitmapId per PinType.

**Comment by inforithmics** [comment]
I found the relevant checkins. 

In this checkin I had put the SkSvg Image into the cache so that it is only created once. Which improved the Performance a lot
https://github.com/Mapsui/Mapsui/pull/981/commits/5034bc6d551eaf573e78b42982510ade9e97fcae

But in this checkin I had to revert it, because it interfered with the Xaml Rendering Backend which doesn't now anything about SkSvg Or SkImage. 
https://github.com/Mapsui/Mapsui/pull/981/commits/5c87e81acb7079622a8e1eb9d16cd0fe9005390e

So To work with the Xaml Rendering Backend I had to cache the MemoryStream (still better than before but slower).

**Comment by janusw** [comment]
> I found the relevant checkins.
> 
> In this checkin I had put the SkSvg Image into the cache so that it is only created once. Which improved the Performance a lot
> [5034bc6](https://github.com/Mapsui/Mapsui/commit/5034bc6d551eaf573e78b42982510ade9e97fcae)
> 
> But in this checkin I had to revert it, because it interfered with the Xaml Rendering Backend which doesn't now anything about SkSvg Or SkImage.
> [5c87e81](https://github.com/Mapsui/Mapsui/commit/5c87e81acb7079622a8e1eb9d16cd0fe9005390e)
> 
> So To work with the Xaml Rendering Backend I had to cache the MemoryStream (still better than before but slower).

Thanks for the comments. That sounds a lot like we need a smarter BitmapRegistry, whose caching mechanism should depend on the Rendering Backend ...?

**Comment by charlenni** [comment]
For me, it is a two part problem: one part is in Mapsui implementation, the other is in Mapsui.Forms.

In Mapsui we have the BitmapRegistry, which should register any type of bitmap, images, what ever you want, before drawing. It should check, if it has a valid format, so that this is correct when drawing. It is a lightweight and easy part. There is also the place for improvements. One would be to convert the string with drawing commands from an SVG to an easy drawable format. This is done by the SKSVG class as you could see [here](https://github.com/wieslawsoltes/Svg.Skia/blob/master/src/Svg.Skia/SKSvg.cs). It is converted in a SKPicture, a very fast format. So the suggestion from @inforithmics is correct: if you add many independent SVGs, then the time for converting them is high, but the drawing is fast. If you create one or hundreds of the same SVG is no problem for drawing time. So I would say, BitmapRegistry is correct, and we could not make things faster.

What could be made better in BitmapRegistry is a check, if an image already exists. But do we want to make this? We have to search to the already registered images and look, if we found the new one. It is time-consuming and not very effective. But could be made (check for length and only compare those images, which having the same length).

The other part is Mapsui.Forms. There is a Pin object, which could have an SVG image. Because this should have changeable colors, I converted this SVG to a bitmap, so that I could use the function to replace colors. It would be nicer (in terms of drawing quality), to use a SKPicture for that, but a long time ago I didn't find a function for this. And because each pin could have another color, I didn't implement a caching algorithm. Wasting time for the normal case. And to be sure, the normal case is not to have more than thousands of pins ;-)

So where should we make changes to support normal cases for everyone? Perhaps to make BitmapId of Pin public would be the best. With this, you could handle images by your own. The setter of BitmapId could check, if the PinType is PinType.Own and if yes, allows to set the intern _bitmapId.

And as you could read in the last comment of inforithmics, there was always a problem with the Xaml renderer. In an ideal world (only one renderer system), you could make improvements, that are not possible with two or more.

**Comment by charlenni** [comment]
Very strange writing. English isn't my preferred language ;-)

To be clear: BitmapRegistry is renderer independent and doesn't know anything about the registered image formats. There exists no SKPicture or SKImages. Forget all SkiaSharp things in this part of code.

**Comment by tmijieux** [comment]
I just was just recently interested in such performance concerns with the Mapsui.Forms mapview(I eventually found out most of my performance issues where coming from another piece of my application but I can still share my thoughts about it),


One thing I noticed (but didn't investigate much further) is that the pins in the MapView use an `private readonly ObservableRangeCollection _pins` and is exposed as an `public IList<Pin> => _pins`. Wouldn't calling `Add()` on the Pins property call the CollectionChanged event handler (and thus the `Refresh()` itself calling the `RefreshData()` method) each time we add a pin ? (this is a real question, I did not tested it...)


> Thanks for the comments. That sounds a lot like we need a smarter BitmapRegistry, whose caching mechanism should depend on the Rendering Backend ...?

@janusw just like @inforithmics said the `SymbolCache` class in the SkiaRenderer already does exactly that.

> Actually I don't fully understand what's the point of using this BitmapRegistry at all, if it does not even help to deduplicate the bitmap references. Why not use references to the bitmap data directly instead of generating integer IDs for them?

Generating and integer ID for them is probably the sole purpose of this class, deduplication is the responsibility of the caller,
the problem actually lies in the MapView Pins implementation which we could probably expects to do that since it is a rather high level component. That integer ID is used in the renderer specific cache `SymbolCache` (at least in the skia renderer, I did nto read about the xaml renderer) to identify the resource.


> Why not use references to the bitmap data directly

Technically it could probably be possible to store opaque `object` in the BitmapRegistry that actually are `SKImage, SKPicture` or the Xaml Renderer counterparts, created in a renderer specific factory, ... but i do not think it has any advantage over the current implementation other than saving a little of memory (maybe ?), and it may probably make the code more complicated to think about.

I figured I could just use the mapsui api directly as i didn't required much of the Forms.MapView functionnality from the beginning. 

Here is the code i ended up writing for my use case. Here I rasterized the SVGs, which may not be what you want if you need variable size SVGs, the end result is not perfect either from an aliasing point of view, I probably have missed something here.
```csharp
using Mapsui.Geometries;
using Mapsui.Layers;
using Mapsui.Providers;
using Mapsui.Styles;
using Mapsui.UI.Forms;
using Mapsui.UI.Objects;
using MyApp.Helpers;
using SkiaSharp;
using Svg.Skia;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace MyApp.Controls
{
    public class PinBitmapHelper
    {
        private int CreateBitmap(Stream data, double scale)
        {
            var svg = new SKSvg();
            svg.Load(data);

            var info = new SKImageInfo(
                (int)(svg.Picture.CullRect.Width * scale),
                (int)(svg.Picture.CullRect.Height * scale)) {
                AlphaType = SKAlphaType.Premul
            };

            var bitmap = new SKBitmap(info);
            var canvas = new SKCanvas(bitmap);
            canvas.Clear();
            canvas.Scale((float)scale);

            // paint.ColorFilter = SKColorFilter.CreateBlendMode(
            //     Color.ToSKColor(),
            //     SKBlendMode.SrcIn
            // ); // use the source color

            using var paint = new SKPaint() { IsAntialias = true };
            canvas.DrawPicture(svg.Picture, paint);

            using var image = SKImage.FromBitmap(bitmap);
            using var bitmapData = image.Encode(SKEncodedImageFormat.Png, 100);
            var bitmapDataArray = bitmapData.ToArray();
            var stream = new MemoryStream(bitmapDataArray);
            var bitmapId = BitmapRegistry.Instance.Register(stream);
            return bitmapId;
        }

        private readonly Dictionary<string,int> resToBitmapId = new Dictionary<string,int>();

        public int GetOrCreateBitmapFromSvgData(string name, string svgData, double scale)
        {
            if (!name.StartsWith("data://"))
            {
                name = "data://" + name;
            }
            if (resToBitmapId.TryGetValue(name, out int bitmapId))
            {
                return bitmapId;
            }
            var data = new MemoryStream(Encoding.UTF8.GetBytes(svgData));
            bitmapId = CreateBitmap(data, scale);
            resToBitmapId[name] = bitmapId;
            return bitmapId;
        }

        public int GetOrCreateBitmapFromSvgResource(string resName, double scale)
        {
            if (resToBitmapId.TryGetValue(resName, out int bitmapId))
            {
                return bitmapId;
            }
            string shortName = resName.Replace("resource://MyApp.Assets.","");
            string svgData = AssetFileLoader.LoadFile(shortName);
            var data = new MemoryStream(Encoding.UTF8.GetBytes(svgData));
            bitmapId = CreateBitmap(data, scale);
            resToBitmapId[resName] = bitmapId;
            return bitmapId;
        }
    }

    public class MyPin : IFeatureProvider
    {
        private static readonly PinBitmapHelper pinHelper = new PinBitmapHelper();
        public Feature Feature { get; private set; }
        public Position Position { get; }

        public MyPin(double x, double y,
                     double offsetX=0.0, double offsetY=0.0,
                     double scale=1.0,
                     string color="#FFFFFF")
        {
            Position = new Position(y, x);
            Feature = new Feature {
                Geometry = Position.ToMapsui(),
                Styles = new List<IStyle> {
                    new SymbolStyle {
                        SymbolType = SymbolType.Ellipse,
                        Fill = new Brush { Color=Color.FromString(color) },
                        Outline = new Pen { Color = Color.Transparent },
                        SymbolOffset = new Offset(offsetX, offsetY),
                        SymbolScale=scale,
                    }
                }
            };
        }

        public MyPin(string resName,
                     double x, double y,
                     double offsetX=0.0, double offsetY=0.0,
                     double scale=1.0, string svgData=null)
        {
            Position = new Position(y, x);

            int bitmapId;
            if (svgData != null)
                bitmapId = pinHelper.GetOrCreateBitmapFromSvgData(resName, svgData, scale);
            else
                bitmapId = pinHelper.GetOrCreateBitmapFromSvgResource(resName, scale);

            Feature = new Feature {
                    Geometry = Position.ToMapsui(),
                    Styles = new List<IStyle> {
                    new ImageStyle {
                        BitmapId = bitmapId,
                        SymbolScale = 1.0,
                        SymbolOffset = new Offset(offsetX, offsetY),
                    }
                }
            };
        }
    }

    public class MapView : MapControl
    {
        private readonly MemoryLayer _mapPinLayer;
        private readonly MemoryLayer _locationLayer;
        private List<MyPin> _myPins;
        public event EventHandler<MyPin> PinClicked;
        private static int bitmapStillId = -1;


        private MemoryLayer CreateLocationLayer()
        {
            if (bitmapStillId == -1)
            {
                var bitmapStill = Mapsui
                    .Utilities
                    .EmbeddedResourceLoader
                    .Load($"Images.MyLocationStill.svg", typeof(MyLocationLayer));

                if (bitmapStill != null)
                {
                    // Register bitmap
                    bitmapStillId = BitmapRegistry.Instance.Register(bitmapStill);
                }
            }
            var defaultPos = new Position(GpsPosition.DefaultPosition.Latitude,
                                          GpsPosition.DefaultPosition.Longitude);
            var layer = new MemoryLayer("locationLayer") {

                Style = null,
                DataSource = new MemoryProvider(new List<IFeature> {
                    new Feature {
                        Geometry = defaultPos.ToMapsui(),
                        Styles = new List<IStyle> {
                            new ImageStyle {
                                BitmapId = bitmapStillId,
                            }
                        }
                    }
                }),
                Enabled = false,
            };
            return layer;
        }

        public void SetMyLocation(Position myPosition)
        {
            var provider = _locationLayer.DataSource as MemoryProvider;
            provider.Features.First().Geometry = myPosition.ToMapsui();
            if (MyLocationFollow)
                MoveToMyLocation();
            Refresh();
        }

        public void MoveToMyLocation()
        {
            var provider = _locationLayer.DataSource as MemoryProvider;
            var point = provider.Features.First().Geometry as Point;
            Navigator.CenterOn(point);
        }

        public MapView()
        {
            _mapPinLayer = new MemoryLayer("pinLayer") {
                Style = null,
                DataSource = new MemoryProvider(new List<IFeature>()),
                IsMapInfoLayer = true,
            };
            _locationLayer = CreateLocationLayer();
            TouchMove += (s, e) => { MyLocationFollow = false; };

            SetPins(new List<MyPin>(), refresh: false);
            Map.Layers.Add(_mapPinLayer);
            Map.Layers.Add(_locationLayer);
            Info += (s,e) => {
                if (_myPins == null)
                    return;
                if (e.MapInfo.Layer != _mapPinLayer)
                    return;
                foreach (var pin in _myPins)
                {
                    if (pin.Feature.Equals(e.MapInfo.Feature))
                    {
                        PinClicked?.Invoke(this, pin);
                        break;
                    }
                }
            };
        }

        public bool MyLocationEnabled
        {
            get => _locationLayer.Enabled;
            set => _locationLayer.Enabled = value;
        }

        private bool _myLocationFollow;
        public bool MyLocationFollow
        {
            get => _myLocationFollow;
            set
            {
                _myLocationFollow = value;
                if (value && MyLocationEnabled)
                {
                    MoveToMyLocation();
                }
            }
        }

        public void SetPins(List<MyPin> myPins, bool refresh=true)
        {
            if (myPins == null)
                myPins = new List<MyPin>();
            _myPins = myPins;

            var demPins = myPins.Select(p => p.Feature).ToList();
            _mapPinLayer.DataSource = new MemoryProvider(demPins);
            if (refresh)
            {
                Refresh(); // this call InvalidateSurface();
            }
        }
    }
}

```

**Comment by janusw** [comment]
> I found the relevant checkins.
> 
> In this checkin I had put the SkSvg Image into the cache so that it is only created once. Which improved the Performance a lot
> [5034bc6](https://github.com/Mapsui/Mapsui/commit/5034bc6d551eaf573e78b42982510ade9e97fcae)
> 
> But in this checkin I had to revert it, because it interfered with the Xaml Rendering Backend which doesn't now anything about SkSvg Or SkImage.
> [5c87e81](https://github.com/Mapsui/Mapsui/commit/5c87e81acb7079622a8e1eb9d16cd0fe9005390e)
> 
> So To work with the Xaml Rendering Backend I had to cache the MemoryStream (still better than before but slower).

I tried to re-enable the SkSvg optimization from the first commit (with the patch below), but I see no strong performance improvement over the status quo of the SvgSample (it is still significantly slower than the ManyPinsSample).

So, if I care for performance, should I stick to bitmap images, or do you think it would be possible to get close to bitmap performance with SkSvg?

```
-------------------- Mapsui.Rendering.Skia/BitmapHelper.cs --------------------
index 23036e19..d9d572f7 100644
@@ -29,6 +29,16 @@ namespace Mapsui.Rendering.Skia
                 return new BitmapInfo {Bitmap = image};
             }
 
+            if (bitmapStream is SKSvg skSvg)
+            {
+                return new BitmapInfo { Svg = skSvg };
+            }
+
+            if (bitmapStream is SKImage skImage)
+            {
+                return new BitmapInfo { Bitmap = skImage };
+            }
+
             if (bitmapStream is Sprite sprite)
             {
                 return new BitmapInfo {Sprite = sprite};

--------------- Samples/Mapsui.Samples.Common/Maps/SvgSample.cs ---------------
index 777c4c74..ef4f0339 100644
@@ -10,6 +10,7 @@ using Mapsui.Samples.Common.Helpers;
 using Mapsui.Styles;
 using Mapsui.UI;
 using Mapsui.Utilities;
+using Svg.Skia;
 
 namespace Mapsui.Samples.Common.Maps
 {
@@ -78,7 +79,9 @@ namespace Mapsui.Samples.Common.Maps
                 {
                     var assembly = typeof(PointsSample).GetTypeInfo().Assembly;
                     var image = assembly.GetManifestResourceStream(imagePath);
-                    id = BitmapRegistry.Instance.Register(image);
+                    var svg = new SKSvg();
+                    svg.Load(image);
+                    id = BitmapRegistry.Instance.Register(svg);
                     imageCache[imagePath] = id;
                 }
                 catch (Exception exception)

```

**Comment by pauldendulk** [comment]
> So then, with this in mind, wouldn't it be a good idea to add a new pin type (maybe called PinType.CachedSvg), which would take SVG data as input, but cache it into a bitmap for performance reasons (as for PinType.Pin, but with a freely chosen svg as input)? If you agree, I'd volunteer to implement that.

For performance the best solution would be to store bitmaps in the BitmapRegistry. Does the BitmapRegistry need to know if that bitmap originates from an SVG? Why not convert the SVG to bitmap early on and treat it like a bitmap after that? Perhaps render quality? Also an SVG could be rendered with different colors, but this could be the responsibility of a separate layer which generates different images and imageIds for the same SVG with different colors.

A general thing. As charlenni mentioned, the Forms Pins are a convenience layer. If the focus is on performance perhaps Forms should be avoided altogether. Also, in November this year MAIU is planned to be released. So better to focus on platform independent rendering.

**Comment by janusw** [comment]
> > So then, with this in mind, wouldn't it be a good idea to add a new pin type (maybe called PinType.CachedSvg), which would take SVG data as input, but cache it into a bitmap for performance reasons (as for PinType.Pin, but with a freely chosen svg as input)? If you agree, I'd volunteer to implement that.
> 
> For performance the best solution would be to store bitmaps in the BitmapRegistry.

Yes, this currently looks like the best option to me.

> Does the BitmapRegistry need to know if that bitmap originates from an SVG?

I don't think it does.

> Why not convert the SVG to bitmap early on and treat it like a bitmap after that? Perhaps render quality?

Sure, one can do the conversion on the user side (and I currently do that as a workaround). But I still think it would be nice to have an option in Mapsui to do it automatically. It would really only need a new PinType, the conversion code is already there and with a bit of documentation it will save many users a lot of time to find out: Whoops, svg pins are creepily slow.

The new PinType would also function as an implementation suggestion that basically says:

> Dear user, you can use both svg pins and bitmap pins in Mapsui, but if you choose svg, please be aware that it might be very slow if you have many pins. In case you do have svg pins, we can conveniently buffer them into a bitmap for better performance. The drawing quality might be slightly worse, though. Choose whatever you prefer.

I have svg icons that I also use in non-map parts of my app (where the performance is no problem), so I'd like to keep them in my assembly without adding a png version for each one.

> A general thing. As charlenni mentioned, the Forms Pins are a convenience layer. If the focus is on performance perhaps Forms should be avoided altogether.

Nope, strongly disagree here for a couple of reasons:

1. All of Mapsui is a convenience layer. In principle any user can re-implement all of it on his own. No one wants to do that. People want a nice API for displaying maps without going into all the gory details.
2. I'll not be moving away from Mapsui.Forms. I need a cross-platform implementation, and I don't want to implement things for each platform separately.
3. Using a MapControl instead of a MapView might be slightly more feasible, but even here: I think it's a useful high-level abstraction. I don't want to rewrite all the boiler-plate code (and I assume most other users don't want that either).
4. It's a fallancy to think that everything will be automatically more performant if one uses the lower-level APIs or platform specific-code. See my comment above about how SvgSample (which is not using Mapsui.Forms) is much slower than ManyPinsSample (which is using Mapsui.Forms and the Pins class). The latter is blazingly fast on any hardware I tested, even with a large number of pins. The big difference is whether you use svg icons or bitmap icons.
5. Synergy effects: If every user implements their own version of a MapView, everyone will have their own bugs in there (and possibly bad design decisions). We as a Mapsui community should provide a MapView that is as performant and convenient as possible, with all performance bottlenecks identified and removed.

If it weren't for Mapsui.Forms and high-level components like a MapView with Pins and Callouts, I wouldn't be using Mapsui at all.

**Comment by pauldendulk** [comment]
> If the focus is on performance perhaps Forms should be avoided altogether.

I did not mean to say Mapsui.Forms should be avoided altogether, but that it should be avoided with respect to  SVG Pins if you need high performance. And a Forms independent alternative can still be user friendly.

**Comment by janusw** [comment]
> > If the focus is on performance perhaps Forms should be avoided altogether.
> 
> I did not mean to say Mapsui.Forms should be avoided altogether, but that it should be avoided with respect to SVG Pins if you need high performance.

Well, the lesson that I take from the discussion in this issue so far is that SVG pins are to be avoided, if you need performance. Doesn't matter if you use them through Mapsui.Forms or not.

Again:
* ManyPinsSample is fast (based on Mapsui.Forms, but without SVGs).
* SvgSample is slow (not based on Forms, but with SVGs).

> And a Forms independent alternative can still be user friendly.

Might be. But it requires you to write platform-specific code, right?

(Btw, we're not talking "end-user friendly" here, but "developer friendly", right?)

Platform-specific code is not what I call developer friendly. It's one of the great achievements of the Xamarin.Forms framework that you can define all your UI and logic in a platform-independent way, which saves a lot of development and maintenance time. As long as you're developing for one specific platform, that may not matter to you, but if you develop a large app framework that targets multiple platforms, it's a crucial advantage.

**Comment by pauldendulk** [comment]
> (Btw, we're not talking "end-user friendly" here, but "developer friendly", right?)

Yes, developer friendly.

Note, there might be some confusion in terminology because I see Xamarin.Forms as just one platform within Mapsui but Xamarin.Forms itself supports multiple platforms.

In general I would like to see more functionality in Mapsui core that is now still only in Mapsui.Forms. Putting a pin on a map or showing a dialog does not have to depend on Forms.

**Comment by janusw** [comment]
> Note, there might be some confusion in terminology because I see Xamarin.Forms as just one platform within Mapsui but Xamarin.Forms itself supports multiple platforms.

Well, ok. When I used the word 'platform' above, I was basically referring to the OS you develop for. Xamarin.Forms supports things like Android, iOS, MacOS, Windows (via UWP or WPF), Linux (via GTK#). To XF itself I would rather refer as a 'framework' than a 'platform', but let's not start splitting hairs here (as long as we can agree on a common language).

 
> In general I would like to see more functionality in Mapsui core that is now still only in Mapsui.Forms. Putting a pin on a map or showing a dialog does not have to depend on Forms.

I see this from a somewhat different angle: As far as I am concerned, Mapsui.Forms is the only relevant 'platform' ('backend'?) for Mapsui. Doesn't it sort of supersede all the others, since it supports all the OS platforms that Mapsui also supports 'natively' (Android, iOS and Windows via UWP and WPF) and even more (MacOS and possibly Linux/GTK#)?

Why would I ever use Mapsui.UI.Android, if I can instead use Mapsui.UI.Forms and get all other platforms "for free" instantly?

**Comment by charlenni** [comment]
@janusw That's a developer decision. There are reasons, why someone wants to develop only for Android or only for WPF.

We were, before the discussion goes in a platform direction, at the point, why SVG pins slow, while bitmap pins are fast. There could be two reasons:
1. Conversion from SVG to drawing commands is slow
2. Drawing the commands to the canvas is slow
To check this, we have to make some performance checks.

One problem I had, was, that Mapsui.Forms relays on SkiaSharp, but the underlying Mapsui doesn't. The used renderer than use SkiaSharp again. So I had to bring special SkiaSharp objects through the Mapsui part, so that they are usable. I made this by changing the `Register(Stream data)` function to `Register(object data)`. With this, it is possible to bring all types of SkiaSharp objects into the renderer.

Bad is, that for an easy way, we have to change the renderer to handle this object.

Ok, that are the circumstances.

Now to the pin problem. As I mentioned before, I didn't find a way to change the color of an SVG pin. If you have better knowledge, perhaps you have a way to do this. My solution was to convert the SVG pin to a bitmap pin and change the color for this bitmap. That was the reason, why I didn't implement a caching, because the user could change the color of the pin, and we need a new bitmap for each color. An Advantage of this is, that the pin draws faster, because copying bitmaps is really fast. The Disadvantage is, that the bitmap isn't as sharp as the SVG.

So you have to decide, what you want:
1. Accept lesser quality and use bitmaps. Works already.
2. Want to have better quality, than there are two ways:
a. Add a new PinType for this and extend Mapsui.Rendering.Skia to render this correct.
b. Add your own layer with your own SytleRenderer and draw your pins by yourself. 

Sorry, that I could not say, that this works out of the box.

**Comment by charlenni** [comment]
BTW. is it possible to see, what your pins look like?

**Comment by janusw** [comment]
> BTW. is it possible to see, what your pins look like?

I'm afraid I cannot share them here. Is it relevant how they look like? They're not even horribly large or complicated. Typically around 1kB in size.

Mapsui's SvgSample demonstrates nicely that already rather small SVGs like the Pin.svg in the Mapsui repo generate a considerable performance hit (if you have a large number of them).

**Comment by charlenni** [comment]
It is relevant in the case of the complexity of the pin. If you have to draw this pin each time, it could be a problem.

**Comment by charlenni** [comment]
Ok, when I look into the code, then there some "special" problems.

The conversion from string to stream is only for the old version of `BitmapRegistry`. The new one, which use `object` doesn't need such conversion. As you could see, the stream is consumed by module BitmapHelper in Skia renderer (see [here](https://github.com/Mapsui/Mapsui/blob/b4df2b9388cc94d61f54a09fea7f99967a392e54/Mapsui.Rendering.Skia/BitmapHelper.cs#L13)) or in Xaml renderer (see [here](https://github.com/Mapsui/Mapsui/blob/b4df2b9388cc94d61f54a09fea7f99967a392e54/Mapsui.Rendering.Xaml/SymbolCache.cs#L13)). But we need no stream. We could provide a string. That's enough. 

So the best way would be to provide a string instead of a stream. Or in the case of the Skia renderer, we could provide a ready drawn SKPicture, which is much faster than a non converted string.

But we can't change the Forms Pin, because of the color. Or do you have any idea?

**Comment by tmijieux** [comment]
I think the SvgSample re-use always the same 1 "bitmapId" ([here](https://github.com/Mapsui/Mapsui/blob/master/Samples/Mapsui.Samples.Common/Maps/SvgSample.cs#L73-L92)). During rendering the Svg is parsed the first time but the second time it is already parsed as a SKPicture and stored in the SymbolSache ([here](https://github.com/Mapsui/Mapsui/blob/b4df2b9388cc94d61f54a09fea7f99967a392e54/Mapsui.Rendering.Skia/SymbolCache.cs#L10-L14) (skia)or [here](https://github.com/Mapsui/Mapsui/blob/b4df2b9388cc94d61f54a09fea7f99967a392e54/Mapsui.Rendering.Xaml/SymbolCache.cs#L13-L52) (xaml)) so the problem is not with the Svg parsing since it is already done once,(on xaml there is a call to the Freeze() method which according to the comment if i understand correctly it even rasterize it ?) 

I dont think it will be "much faster" to do it on the BitmapRegistry.
 
Also questioning the "quality" problem, It seems weird to me that we can't achieve the same level of quality we rasterizing an svg than when we draw the svg using the parse SKPicture, is it not possible that the problem is just with the SKBitmap dimension are not correct to match the screen density and is scaled up afterwards ?

**Comment by inforithmics** [comment]
Here additional Information:

If you create your Svgs with Adobe Illustrator they are where slow, when you clean up your Svg. They render parse much faster afterwards.

Program to clean up Svg.
https://github.com/RazrFalcon/svgcleaner-gui

Gitthub issue:
https://github.com/svg-net/SVG/issues/784

**Comment by charlenni** [comment]
While working on #1091 I tested the Many Pin Sample. I don't know why (any HTTP error), but I had no map in the background. With this, the sample is "very" fast. No hanging. So I assume, that drawing the map is the problem and not the Pins.

**Comment by janusw** [comment]
> While working on #1091 I tested the Many Pin Sample. I don't know why (any HTTP error), but I had no map in the background.

That's strange. The OSM background map works well for me in this sample (both on iOS and Android).


> With this, the sample is "very" fast. No hanging.

Well, yes. The ManyPinsSample is fast also for me. The SvgSample is the one that is slow (in particular on iOS / iPhone8).


> So I assume, that drawing the map is the problem and not the Pins.

I don't think this is the right conclusion.

**Comment by charlenni** [comment]
While experimenting with my demo project I encountered a problem, that is perhaps the reason for the performance problem of Mapsui.

In the demo project I refresh the screen each 1/60 s if needed. This is done by `InvalidateSurface()`. Up to now, I called this always on the main thread. If I look, how long this takes, I get the following results:

> Last draw needs 0.160 ms
Last draw needs 0.168 ms
Last draw needs 0.133 ms
Last draw needs 0.216 ms
Last draw needs 0.200 ms
Last draw needs 0.198 ms

The main thread is blocked for around 200 ms while drawing.

If I call `InvalidateSurface()` from the timer thread (non main thread), then I get the following results for the blocking of the main thread:

> Last draw needs 0.005 ms
Last draw needs 0.005 ms
Last draw needs 0.005 ms
Last draw needs 0.005 ms
Last draw needs 0.004 ms
Last draw needs 0.005 ms

No differences in the result, but is 40x faster back in the main thread.

With these results, I would say, SkiaSharp could draw in the background, if there is no access to main thread parts while drawing. This isn't used for Mapsui. Mapsui is much slower than it should be. The control isn't very reactive.

**Comment by janusw** [comment]
> No differences in the result, but is 40x faster back in the main thread.

Wow, that's an interesting observation. It might also explain a possible platform dependence of this phenomenon (due to different main-thread handling?). Generally I observe Mapsui-based apps to have rather good performance on Android, but in some cases very bad performance on iOS.
 
> With these results, I would say, SkiaSharp could draw in the background, if there is no access to main thread parts while drawing. This isn't used for Mapsui.

Would be great if this worked. Could you push your InvalidateSurface change to a branch, so that one can try it?

> Mapsui is much slower than it should be. The control isn't very reactive.

Yes, the control being unresponsive to user interaction is really the main problem I'm seeing ...

**Comment by charlenni** [comment]
@janusw The problem is, that it at the moment isn't working with Mapsui. Perhaps there is some part, that access a part, which only can be access by the main thread. It seems much more work to do this.

Meanwhile, I added some vector tiles to my [demo](https://github.com/charlenni/VTMap). It is a MBTiles file of Monaco. There isn't any work on performance improvement up to now, so I think, the performance could be made even better.

**Comment by tmijieux** [comment]
```csharp
if (redraw)
{
    dt = _stopWatch.Elapsed.TotalMilliseconds;
    Invalidate();
    Debug.WriteLine($"Last draw needs {(_stopWatch.Elapsed.TotalMilliseconds-dt).ToString("N3", CultureInfo.InvariantCulture)} ms");
}

```
The actual drawing is not necessarily done synchronously during the `InvalidateSurface()` call, the actual draw may not even be done on the UI thread, if you want to measure the duration of a draw you must time at the beginning and the ending of the PaintSurface event handler. Look at [this test project I made](https://github.com/tmijieux/TestForMapsui), the actual number of draws is not equal to the number of call of InvalidateSurface() and (at least on my phone) the PaintSurface is not called on the main thread at all, and the PaintSurface method may be called after InvalidateSurface returns ( the conditions when this happens in that particular project may be biased because my PaintSurface is doing almost nothing except clearing the screen)

also the documentation clearly states that invalidateSurface must be called on UI thread

https://docs.microsoft.com/en-us/dotnet/api/skiasharp.views.forms.skcanvasview.invalidatesurface?view=skiasharp-views-forms-2.80.2
`This needs to be called from the main thread.`

**Comment by charlenni** [comment]
Above you mentioned the documentation from Microsoft regarding SKCanvasView. That is correct. If you look into [SKGLView](https://docs.microsoft.com/en-us/dotnet/api/skiasharp.views.forms.skglview.invalidatesurface?view=skiasharp-views-forms-2.80.2), you won't find such a comment. So SKCanvasView and SKGLView are different. 

So the measurement above will show, how long the `Invalidate()` uses on this (main?) thread.

I would say, SKGLView does the work in the background and needs only a small-time amount on the main thread for copying the screen. SKCanvasView needs much more time on the main thread. That's what I see on my vector tile demo, which never calls `InvalidateSurface()` on the main thread.

And yes, a benchmarking version would be nice.

**Comment by tmijieux** [comment]
> Above you mentioned the documentation from Microsoft regarding SKCanvasView. That is correct. If you look into SKGLView, you won't find such a comment. So SKCanvasView and SKGLView are different.

this may also be just something that was forgotten , there is a lot of "To be added" across the documentation of Xamarin.Forms / Skiasharp so I would not rely on the quality of the documentation to take that conclusion.

> I would say, SKGLView does the work in the background and needs only a small-time amount on the main thread for copying the screen. SKCanvasView needs much more time on the main thread. That's what I see on my vector tile demo, which never calls `InvalidateSurface()` on the main thread.

if you replace the SKGLView by a SKCanvasView in my example, you can see that most of the conclusion remains the same,
like you said the drawing is done on the main thread, with SKCanvasView but even if you use the button that calls InvalidateSurface() 300 times, you can see that paintsurface is only called once after the ButtonCLick handler returns
so the implementation of InvalidateSurface is just most probably setting a boolean to indicate a draw need to be done, and probably like i said there is maybe a timer that will trigger the draw immediately during the InvalidateSurface() if the time that has passed since last draw is greater than a certain offset. I suspect this is what is happening in your case since your draw is very long, the timer is always exhausted when you call InvalidateSurface(), but that does not change the fact that you should measure in the PaintSurface event handler to get the timing reliably

> So the measurement above will show, how long the Invalidate() uses on this (main?) thread.

just exactly that, nothing less, nothing more, the Invalidate call MAY OR MAY NOT  call the drawing synchronously

**Comment by charlenni** [comment]
I assume, that this wasn’t forgotten. It doesn‘t make sense to draw things for a special GPU on the main thread. So all things for the SKGLView should be drawn on an extra thread. For SKCanvasView this is a totally different thing. It is CPU base and there are things, that have to run on the main thread.

**Comment by charlenni** [comment]
I don‘t want to measure the time the drawing needs on another thread. I‘m interested in the time, I use on the main thread. This keeps Mapsui reactive.

**Comment by tmijieux** [comment]
> It doesn‘t make sense to draw things for a special GPU on the main thread

what do you mean exactly ? you can control gpus from any thread/core.

To have mapsui responsive, you need to draw 1 frame every 16 ms (to achieve 60fps), so the thread it is drawn on does not matter.
Not doing things on the main thread does not automagically make things responsive, the opposite only is true only : doing blocking operations on the main thread may make things unresponsive.
If that SKGLView thread is producing 1 frame every 200ms you will see only one frame every 200ms.


The main thread can have additional worker threads to perform the drawing because it needs to render multiple things and doing them in parallel may improve performance if you have multiple cores, but the framework still cant allow user code to modify the UI data while drawing is going on so there will probably be synchronisation to wait for all drawing threads to finish their tasks before handling back control to user code. (i.e. if you modify the styles while that SKGLView thread is drawing, that would still be a bug, i assume during the time SKGLView is drawing the main thread is free to draw other UI(for example other xamarin forms) components, but not user code)

There is a very easy experience we can make to verify this : is to add a blocking 5seconds sleep in PaintSurface (on the SKGLView thread) and see if the rest of the interface is still responsive...

The result of my experience is, ... well the main thread does not wait for the SKGLView paint to complete and if you update the ui during the SKGLView is drawing, then you can have race conditions in the drawing thread, it seems very much like a bug to me

see that issue other people are asking about this: https://github.com/mono/SkiaSharp/issues/1465#issuecomment-799277533
(seems the extra thread to draw SKGLView is only on android)

**Comment by charlenni** [comment]
So you will say, that you like to wait, say e.g. 200 ms, to draw the map on the main thread, while nothing else could be done. Correct?

**Comment by tmijieux** [comment]
that is the whole point of the UI thread , (which is "nothing else can be done on the UI data hierarchy while you draw, the ui thread mechanism purpose is to serialize all access to ui data, so that there is no race conditions between update and drawing). What i would prefer is obviously for the map to draw very fast.

https://github.com/tmijieux/TestForMapsui/blob/skglview-thread-android/TestMapsui/TestMapsui/MainPage.xaml.cs
in my update example for the SKGLView i added a thread.sleep() in the SKGLView (PaintSurface call) to simulate for a long drawing time, you can see if you test it that if you press any button the program will crash because the ui is updated during the lifetime of the PaintSurface call, that would not happen if the drawing was done, The sleep is exagerated to trigger the problem everytime but this could technically happen without the Thread.Sleep and Task.Delay, the solution i can think of to avoid any concurrent access problem is to have the whole data hierarchy accessed by the SKGLView PaintSurface be hold in an immutable data hierarchy, and an update to the map would replace the whole hierarchy, that way we would have no race problems

**Comment by tmijieux** [comment]
to give an example more closely related to what happen in mapsui imagine the user is updating (on the main ui thread) the collection that is backing the GetFeatures() return value of some layer provider(for example, the user is adding or removing a feature from the collection), at the same time on the SKGLView drawing thread the renderer is iterating over that same collection in the VisibleFeatureIterator , what do you think will happen ?

**Comment by charlenni** [comment]
You are right. That leads to problems. But for such a scenario Mapsui isn't made. In most cases Mapsui is fast enough to draw all in a very short time (bitmap tiles with few features). But if you start to add many features, you see a big loose in performance. With this in mind, Mapsui isn't ready for many Pins or vector tiles. If you know this, all is ok.

**Comment by janusw** [comment]
This can be closed. Most of the performance issues described above were basically fixed for me with #1192.
